primary_beam_correction.py

```diff
@@ -0,0 +1,296 @@
+        vh_beam_mean = np.sqrt(v_beam_rad*h_beam_rad)
+        flux_density = []
+        for nu in c:
+            nu = nu/1.e9
+            theta_b = vh_beam_mean*(1./nu)
+            ratio = rho/theta_b
+            num = np.cos(1.189*np.pi*ratio)
+            dem = 1-4*(1.189*ratio)**2
+            a_b = (num/dem)**2
+            flux_density.append(a_b)
+    return flux_density
+
+
+def mad(data):
+    '''
+    Calculating meadian absolute deviation (MAD).
+    '''
+    MAD_TO_SD = 1.4826
+    med = np.nanmedian(data)
+    dev = np.abs(data - med)
+    return med, MAD_TO_SD * np.nanmedian(dev)
+
+
+def standard_deviation(data):
+    '''
+    Calcalating the standard deviation of each frequency plane using MAD.
+    Returns: Weights per each frequency plane
+    '''
+    diff = 1
+    i = 0
+    data = data[data != 0.0]
+    if len(data) == 0:
+        return float(0.0), float(0.0)
+    med, sd = mad(data)
+    while diff > 0:
+        i += 1
+        if i > 50:
+            return sd
+        old_sd = sd
+        old_len = len(data)
+        cut = np.abs(data - med) < 5.0 * sd
+        if np.all(~cut):
+            return sd
+        data = data[cut]
+        sd = mad(data)[1]
+        diff = old_len - len(data)
+        if sd == 0.0:
+            return old_sd
+    return 1/(sd)**2
+
+
+def weigted_average(arr, weights):
+    '''
+    Computing weighted average of all the frequency planes.
+    '''
+    wt_average = np.average(arr, weights=weights, axis=0)
+    return wt_average
+
+
+def primary_beam_correction(beam_pattern, path):
+    '''
+    Correcting for the primary beam effects.
+    '''
+    raw_image = read_fits(path)
+    raw_img = []
+    try:
+        nterm = raw_image.header['NTERM']
+    except KeyError:
+        nterm = _get_value_from_history('NTERM', raw_image.header)
+    nterm = int(nterm)
+    for j in range(nterm, raw_image.header['NSPEC']+nterm):
+        raw_img.append(np.ravel(raw_image.data[0, j, :, :]))
+    snr = []
+    pbc_image = []
+    for i in range(len((beam_pattern))):
+        # Getting all the values with attenuated flux of less than 10% of the peak
+        ind = np.argwhere(beam_pattern[i] <= 0.10)
+        if len(ind) > 0:
+            beam_pattern[i][ind] = np.nan
+        # Getting the rms noise in each image (before primary beam correction)
+        snr.append(standard_deviation(raw_img[i]))
+        # To correct the effect of the beam we divide by the beam.
+        ratio = raw_img[i] / beam_pattern[i]
+        pbc_image.append(ratio)
+    # Primary beam corrected (pbc) image
+    pbc_image = np.array(pbc_image)
+    snr = np.array(snr)
+    # Calculating a weighted average of the indidvidual frequency lane images
+    corr_image = weigted_average(pbc_image, snr)
+    corr_image = corr_image.reshape(raw_image.data.shape[2], raw_image.data.shape[3])
+    # Add new axis
+    corr_image = corr_image[np.newaxis, np.newaxis, :, :]
```

Comment:
I think you can just do `corr_image = corr_image.reshape(1, 1, raw_image.data.shape[2], raw_image.data.shape[3])`

Though I might be wrong.

Reply:
It is doable and I have done that. Thanks, this will save time for unnecessary computation.